### ISSUE TITLE
[FEATURE] Support for passing timeouts to kubernetes resources

### ISSUE THREAD

**Issue by kahirokunn** [aws-ia/terraform-aws-eks-blueprints#846]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

#### What is the outcome that you are trying to reach?

My network environment is slow, only 64 kbps. 😢
Unfortunately, the Kubernetes cluster to be operated is on the other side of the globe.
It will not complete in 30 seconds, which is very troubling.
Need to extend Kubernetes resources timeout.

Reference: https://github.com/hashicorp/terraform-provider-kubernetes/issues/1755#issuecomment-1174734780

#### Describe the solution you would like

Just support for passing timeouts to Kubernetes resources.

#### Describe alternatives you have considered

I considered modifying the code in the Kubernetes provider, but that seems a bit difficult

#### Additional context

Terraform Cloud is located in Oregon.
The straight line distance from Oregon to Japan is over 5000 miles. It is too far.

<img width="647" alt="スクリーンショット 2022-08-12 8 53 40" src="https://user-images.githubusercontent.com/22343391/184261136-229611ec-903c-432f-842a-e5a3a9dcdfc4.png">

**Comment by kahirokunn** [comment]
> If you are interested in working on this issue or have submitted a pull request, please leave a comment

This is a PR.
https://github.com/aws-ia/terraform-aws-eks-blueprints/pull/734

**Comment by tameto** [comment]
I am encountering the same problem and am having trouble.

**Comment by ogontaro** [comment]
Sometimes timeouts happen in my environment.

**Comment by kouheiadachi** [comment]
I got the same problems.

**Comment by vara-bonthu** [comment]
@kahirokunn I understand your concern but i would like to know more about your issue. Are you encountering this issue only for creating Kubernetes service accounts? 

I can see that Kubernetes provider is using 1 min to 10 min for most of the resources except for `kubernetes_service_account` which is in secs. I think it' s easier to change in Kubernetes provider code to Mins rather than Secs.

**Comment by kahirokunn** [comment]
@vara-bonthu 

There are various resources on the Kubernetes Provider side.

1. I think it would be difficult to treat only kubernetes_service_account specially and change the timeout now.
In fact, the Kubernetes Provider people have stated as much.

https://github.com/hashicorp/terraform-provider-kubernetes/issues/1755#issuecomment-1174750760

2. Also, if the timeout time cannot be specified in the module, the only way is to implement timeout infinity on the Kubernetes Provider side. That is problematic. Is there a best timeout for all users and environments?

3. Many Terraform Providers allow timeouts to be specified on a per-resource basis, and many modules support this.
Wouldn't it be better and more intuitive to specify timeouts on the module side?

**Comment by asami-kawasaki** [comment]
I have the same problem. I would like to solve it.

**Comment by reoring** [comment]
I'm looking forward to resolve this problem as soon as possible.

**Comment by Zvikan** [comment]
Hey all,  please make sure to follow the community notes (see top of the issue):
```
Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
```
And instead:
```
Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
```

@kahirokunn as we continue the discussion about this issue and possible code changes in the repo, I wonder, are you able to have a remote machine located near Oregon where you make any of the Terraform operations (with TF Cloud) there? Something like AWS Cloud9 or a simple EC2 machine.

If you can, I highly recommend you giving it a try.

**Comment by kahirokunn** [comment]
@Zvikan TF Cloud does not allow region changes.
Does that mean I should check if TF Cloud can configure EKS for us-west-2 (Oregon)?

**Comment by bryantbiggs** [comment]
>  Does that mean I should check if TF Cloud can configure EKS for us-west-2 (Oregon)?

Where the EKS cluster is provisioned is controlled by users through the provider block

**Comment by kahirokunn** [comment]
@bryantbiggs @Zvikan 
Sorry, I am not fluent in English.
I am using DeepL.
Could you please tell me what you want me to do and what your intention is, in a language that is easily translated?

**Comment by bryantbiggs** [comment]
You can control where the EKS cluster is created by changing the region within the [AWS provider block](https://registry.terraform.io/providers/hashicorp/aws/latest/docs)

This will create resources in `us-west-2`:
```hcl
provider "aws" {
  region = "us-west-2"
}
```

This will creates resources in `ap-northeast-1`:
provider "aws" {
  region = "ap-northeast-1"
}

**Comment by kahirokunn** [comment]
@bryantbiggs 
Thx.

Since I provide services in Japan, the EKS must reside in ap-northeast-1.
But TF Cloud tries to apply from us-west-2.
To control TF Cloud's execution environment, I have to sign a TF Enterprice contract and pay a huge amount of money.

**Comment by bryantbiggs** [comment]
ok, understood. however, where does this statement come into play then:

> My network environment is slow, only 64 kbps. cry

**Comment by kahirokunn** [comment]
@bryantbiggs 

TF Cloud network speed is unknown.

> My network environment is slow, only 64 kbps.

My parents live in the mountains of Saga, Japan.
I sometimes work with terraform there as well.
This is not TF Cloud related.
The mountains in Saga have very old coaxial cables and the network speed is quite slow.

<img width="1603" alt="スクリーンショット 2022-08-17 22 42 54" src="https://user-images.githubusercontent.com/22343391/185149494-440e8641-39f2-4961-b31b-20dedd77d7e1.png">

**Comment by Zvikan** [comment]
@kahirokunn FYI, we tested [eks-cluster-with-new-vpc](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/examples/eks-cluster-with-new-vpc) using Terraform Cloud and we used `ap-northeast-1` region, and we were not able to reproduce any timeout issues.

Here's what I would do if I were you to help with the slow connection:

1. Create machine (AWS Cloud9 / EC2) in us-west-2 region (according to you, TF Cloud runs there, I was not able to confirm if this is true or not)
2. Setup your terraform cli and cloud using that machine, then, run the normal terraform commands from that machine.

### Why doing it?
At the end of the, Terraform Cloud does the actual API calls from their runners, but if you are using Terraform Cloud with TF CLI, it will still stream the results back to your terminal, there's a chance that due to the low connection you are seeing issues there, you should check the results using the actual web ui itself (https://app.terraform.io/app/.....).

By having a machine closer to TF Cloud, you will the situation explained above, and if you are getting disconnected from your remote machine for any reason, you should be able get back to the session and continue from where you left.

If you still believe there's an issue between TF Cloud and ap-northeast-1 where it times out, please make sure to provide us proper reproducible steps and example that we can use and follow. 

Hope this helps!

**Comment by kahirokunn** [comment]
@Zvikan 

> Create machine (AWS Cloud9 / EC2) in us-west-2 region (according to you, TF Cloud runs there, I was not able to confirm if this is true or not)

It is possible to display TF Cloud's Nat IP with the following code

```
resource "null_resource" "show-terraform-cloud-global-ip" {
  provisioner "local-exec" {
    command = <<EOT
    curl httpbin.org/ip
    EOT
  }
}
```

**Comment by kahirokunn** [comment]
I have tried several times and it is indeed not reproducible.
But we certainly faced this problem before.
It is unclear why it no longer reproduces.
I have created a PR on Kubernetes provider just in case.
https://github.com/hashicorp/terraform-provider-kubernetes/pull/1807

Is it really unnecessary to specify timeout?
Specifying a timeout is more work for the code, but there should be a benefit to the user.
